internal/compiler/expression_tree.rs

```diff
@@ -109,82 +109,89 @@ pub enum BuiltinMacroFunction {
 
 impl BuiltinFunction {
     pub fn ty(&self) -> Type {
+        macro_rules! interned_type {
+            ( $x:expr ) => {{
+                thread_local! {
+                    static TYPE: Type = Type::Function(Rc::new($x));
+                }
+                TYPE.with(|t| t.clone())
```

Comment:
Lots of thread_local... I hope this won't be a problem.

Reply:
I'm also far from happy about this, but I don't see any alternative with Rust, do you? In C++ this would be a simple `static const`, but that's not possible here - yet? If we make these types Sync/Send (with `Arc` instead of `Rc` e.g.) then it would be possible.

That said, I would prefer to do that in a separate PR after this gets in - I believe the `thread_local` approach is fine for now?

Reply:
Something that could make it a bit better would be to merge all thread_local in a single one. Example in https://github.com/slint-ui/slint/blob/e01a121165c24e96dbf37644f9c6cfc319582f18/internal/compiler/typeregister.rs#L80

We could generate this with a macro that would look like


```rust
declare_builtin_functions!{
   #[derive(Debug, Clone, PartialEq, Eq)]
   /// A function built into the run-time
   pub enum BuiltinFunction {
      GetWindowScaleFactor : () -> Type::UnitProduct(vec![(Unit::Phx, 1), (Unit::Px, -1)]),
      Mod : (Type::Int32, Type::Int32) -> Type::Int32,
      // ...          
}
```

(This syntax just being an idea)

Reply:
what about the thread locals I added in `typeregister.rs`, e.g. for `noarg_callback_type`, `strarg_callback_type`, `logical_point_type`, `font_metrics_type` - should I also include those somehow, and merge `BuiltinEnums`, `BuiltinFunctions` and more into one big fat `BuiltinTypes` thread local?

Reply:
That works to some degree, but merging BuiltinFunctions into the BuiltinTypes is not easily doable due to its nested usage of other builtin types (e.g. colorscheme enum, logical point type, ...)

As such, I'll keep this separate - but I'm running into an issue with this approach, as shown in the patch I'll push in a bit - how can I designate between builtin functions that need a `(..)` in their `match` arm? See:

```
   Compiling i-slint-compiler v1.9.0 (/mnt/data1/milian/kdab/slint/internal/compiler)
error[E0532]: expected unit struct, unit variant or constant, found tuple variant `BuiltinFunction::ItemMemberFunction`
   --> internal/compiler/expression_tree.rs:128:23
    |
48  |       ItemMemberFunction(SmolStr),
    |       --------------------------- `BuiltinFunction::ItemMemberFunction` defined here
...
128 |                       $(BuiltinFunction::$Name => Type::Function(self.$Name)),*
    |                         ^^^^^^^^^^^^^^^^^^^^^^ help: use the tuple variant pattern syntax instead: `BuiltinFunction::ItemMemberFunction(_)`
...
135 | / declare_builtin_function_types!(
136 | |     GetWindowScaleFactor: () -> Type::UnitProduct(vec![(Unit::Phx, 1), (Unit::Px, -1)]),
137 | |     GetWindowDefaultFontSize: () -> Type::LogicalLength,
138 | |     AnimationTick: () -> Type::Duration,
...   |
225 | |     UpdateTimers: () -> Type::Void,
226 | | );
    | |_- in this macro invocation
    |
    = note: this error originates in the macro `declare_builtin_function_types` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Reply:
nvm - rubber duck moment, I figured this out now - the syntax for how to use `$()` wasn't obvious to me.

```
macro_rules! declare_builtin_function_types {
    ($( $Name:ident $(($Pattern:tt))? : ($( $Arg:expr ),*) -> $ReturnType:expr $(,)? )*) => {
        #[allow(non_snake_case)]
        pub struct BuiltinFunctionTypes {
            $(pub $Name : Rc<Function>),*
        }
        impl BuiltinFunctionTypes {
            pub fn new() -> Self {
                Self {
                    $($Name : Rc::new(Function{
                        args: vec![$($Arg),*],
                        return_type: $ReturnType,
                    })),*
                }
            }

            pub fn ty(&self, function: &BuiltinFunction) -> Type {
                match function {
                    $(BuiltinFunction::$Name $(($Pattern))? => Type::Function(self.$Name)),*
                }
            }
        }
    };
}
```